SOLR-15705: A deleteById request without _route_ param for compositeid router could be sent to all shard leaders

makosten opened the pull request:
https://issues.apache.org/jira/browse/SOLR-15705

# Description

This is an implementation of distributed delete-by id-when using the CompositeId router with a router field defined and the field value is missing. The current behavior in this circumstance is to ignore delete by id request.

# Solution

This first approach detects this condition and forwards the request to the doDeleteByQuery method. This works, but it would be better to extract the shared code from doDeleteByQuery. I looked at this but need some advice on how to proceed.

# Tests

I've extended an existing test to verify the new behavior.

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `main` branch.
- [x] I have run `./gradlew check`.
- [x] I have added tests for my changes.
- [ ] I have added documentation for the [Reference Guide](https://github.com/apache/solr/tree/main/solr/solr-ref-guide)

epugh commented on solr/core/src/test/org/apache/solr/cloud/FullSolrCloudDistribCmdsTest.java:
```diff
@@ -298,6 +303,29 @@ public void testDeleteByIdCompositeRouterWithRouterField() throws Exception {
+
+      // Tests for distributing delete by id when route is missing from the request
+
+      { // Send a delete request with no route to shard1 for document on shard2, should be distributed
```
This is a "unique" technique, that I haven't seen before in our code base.   I can't really evaluate if it's good or not...   Do we do this elsewhere in the Solr code base?

makosten replied:
In this case I'm copying the style of the existing test I'm modifying. Code blocks outside of flow-of-control are just for variable scoping, so you can include `final UpdateRequest deleteRequest = new UpdateRequest();` repeatedly.

epugh replied:
awesome!  I mis-read the patch file!

epugh commented on solr/core/src/java/org/apache/solr/update/processor/DistributedZkUpdateProcessor.java:
```diff
@@ -325,6 +325,24 @@ protected void doDeleteById(DeleteUpdateCommand cmd) throws IOException {
   @Override
   protected void doDistribDeleteById(DeleteUpdateCommand cmd) throws IOException {
+
+    zkCheck();
```
This is more of a comment, why is this called zkCheck?   icky name.

makosten replied:
It verifies that zookeeper (zk) is available.

epugh replied:
Thanks...   I am tempted to create a JIRA but.....   Will probably just live iwth it.

makosten commented:
I revised the logic to call DeleteByQuery earlier. Now instead of calling doDeleteByQuery frrom doDistribDeleteById, it is called from doDeleteById. I tried revising doDeleteById to forward the request to all shard leaders if the required route is missing from the request, but this did not work. At one time I believe a deleteById would work even if the route was missing if it happened to land on the right shard, but it seems that this behavior is changed.

epugh commented on solr/core/src/java/org/apache/solr/update/processor/DistributedZkUpdateProcessor.java:
```diff
@@ -314,6 +314,23 @@ public void processDelete(DeleteUpdateCommand cmd) throws IOException {
+      if (router instanceof CompositeIdRouter && routeField != null)  {
+        DistribPhase phase = DistribPhase.parseParam(req.getParams().get(DISTRIB_UPDATE_PARAM));
+        if (phase == DistribPhase.NONE) {
+          log.debug("Using compositeId router and deleteById command is with missing route value, distributing to all shard leaders");
```
@makosten small nit, should this debug line be "command is missing" instead of "is with" ;-)

dsmiley commented:
Sorry, I find having `doDeleteById` call `doDeleteByQuery` to be way too kludgy (and you admitted as much in JIRA).  It appears that this works almost accidentally, and is thus fragile and perhaps internally slower than it could be.  It seems the logic should go into `doDistribDeleteById`.

But firstly, I'm confused about how this PR relates to the JIRA issue.  The JIRA issue clearly states this issue occurs when the implicit router is being used.  Yet the test here and the fix use/check-for the CompositeIdRouter.  Am I missing something?

makosten commented:
@cpoerschke You nailed it! The CompositeId Router has some inconsistent behavior. It's strict on the router field when adding but if the route is missing when deleting by id, it returns the slice based on the unique id hash. I updated the CompositeIdRouter to return null for getTargetSlice for this condition, in which case it is run on the current slice, and then flag the request to be broadcast to the other shard leaders.

This also affected the Implicit Router as it will now broadcast the delete-by-id request if the route is missing, so I updated a failing test. I'd like some feedback on this, as maybe the old behavior is more desirable for the Implicit router.

cpoerschke commented:
> This also affected the Implicit Router as it will now broadcast the delete-by-id request if the route is missing, so I updated a failing test. I'd like some feedback on this, as maybe the old behavior is more desirable for the Implicit router.

From the _"... does not automatically route ..."_ mention at https://solr.apache.org/guide/8_10/collection-management.html#create-parameters documentation I think the existing behaviour for the implicit router makes sense i.e. routing of both indexing and delete-by-id requests is the caller's responsibility.

(If we wanted to change that behaviour, hypothetically, then doing so separately (e.g. as a follow-up issue and pull request) might be clearest and non-backwards compatibility would need to be considered and communicated e.g. users currently doing round-robins on delete-by-id external to Solr can cease doing that and users which use the implicit routing to (temporarily?) have the same document id on multiple shards would benefit from a way to opt-out of the delete-by-id broadcasting.)

cpoerschke commented on solr/core/src/java/org/apache/solr/update/processor/DistributedZkUpdateProcessor.java:
```diff
@@ -316,6 +317,41 @@ public void processDelete(DeleteUpdateCommand cmd) throws IOException {
 
+    if (broadcastDeleteById && DistribPhase.parseParam(req.getParams().get(DISTRIB_UPDATE_PARAM)) == DistribPhase.NONE ) {
+
+      log.debug("The deleteById command is missing the required route, broadcasting to leaders of other shards");
```
1/4 - minor: maybe include `cmd.getId()` in the debug logging here

cpoerschke commented on solr/core/src/java/org/apache/solr/update/processor/DistributedZkUpdateProcessor.java:
```diff
@@ -316,6 +317,41 @@ public void processDelete(DeleteUpdateCommand cmd) throws IOException {
+      DocCollection coll = clusterState.getCollection(collection);
+      Collection<Slice> slices = coll.getRouter().getSearchSlices(route, params, coll);
+
+      // if just one slice, we can skip this
```
2/4 - if the "just one slice" logic happened earlier so that `broadcastDeleteById` is only true if there's more than one slice this:
* would remove the need for this `if` here
* could facilitate code sharing with `doDeleteByQuery`

makosten replied:
Another good idea. I moved the check to setupRequest(). I believe the most efficient check that return the same value is coll.getActiveSlicesMap().size().

cpoerschke commented on solr/core/src/java/org/apache/solr/update/processor/DistributedZkUpdateProcessor.java:
```diff
@@ -383,47 +419,7 @@ protected void doDeleteByQuery(DeleteUpdateCommand cmd) throws IOException {
-      outParams.remove("commit"); // this will be distributed from the local commit
-
-      cmdDistrib.distribDelete(cmd, leaders, outParams, false, rollupReplicationTracker, null);
+      boolean leaderForAnyShard = forwardDelete(coll, cmd);
```
3/4 - added a commit to the pull request branch -- feel free to revert or amend -- that factors out `forwardDelete` method from `doDeleteByQuery`, for potential code sharing with `doDeleteById`

makosten replied:
This is great! I'm going to see if forwardDelete can be shared.

makosten replied:
That worked quite nicely!

cpoerschke commented on solr/core/src/java/org/apache/solr/update/processor/DistributedZkUpdateProcessor.java:
```diff
@@ -316,6 +317,41 @@ public void processDelete(DeleteUpdateCommand cmd) throws IOException {
+
     // check if client has requested minimum replication factor information. will set replicationTracker to null if
     // we aren't the leader or subShardLeader
     checkReplicationTracker(cmd);
```
4/4 - `doDeleteByQuery` does rollup replication tracker logic before the forward-delete logic, `doDeleteById` here does check replication tracker logic after the forward-logic -- i've not yet considered this difference in detail, just noticed it whilst factoring out the forward-delete logic

makosten replied:
I added creating the rollup replication tracker before forwarding the delete-by-id to the other shard leaders and saw no behavior difference. The handleReplicationFactor method didn't receive responses from other shard leaders.

cpoerschke reviewed (commented):
Hi @makosten! I haven't yet looked at the test changes but the code changes conceptually make sense to me, comments inline are implementation detail related. What do you think?

makosten commented:
On the unintentional impact to ImplicitIdRouter, I was thinking adding a property to the router as whether to do the broadcast to avoid isinstanceof CompositeIdRouter. But, there are other places in DistributedZkUpdateProcessor that do exactly that.

epugh commented:
Thanks @cpoerschke for digging into this PR ;-)

makosten commented on solr/core/src/java/org/apache/solr/update/processor/DistributedZkUpdateProcessor.java:
```diff
@@ -383,47 +419,7 @@ protected void doDeleteByQuery(DeleteUpdateCommand cmd) throws IOException {
-      outParams.remove("commit"); // this will be distributed from the local commit
-
-      cmdDistrib.distribDelete(cmd, leaders, outParams, false, rollupReplicationTracker, null);
+      boolean leaderForAnyShard = forwardDelete(coll, cmd);
```
Wrapping the log.debug with an isEnabled fixed this.

cpoerschke commented:
Added two commits with small tweaks, yet to complete running full test suite locally but assuming it passes LGTM here.

@makosten - would you like to go add a `solr/CHANGES.txt` entry also?

dsmiley commented on solr/core/src/java/org/apache/solr/update/processor/DistributedZkUpdateProcessor.java:
```diff
@@ -316,6 +317,41 @@ public void processDelete(DeleteUpdateCommand cmd) throws IOException {
   protected void doDeleteById(DeleteUpdateCommand cmd) throws IOException {
     setupRequest(cmd);
```
@sonatype-lift help
